include/recorders.h

```diff
@@ -118,4 +122,333 @@ class command_recorder {
 	std::vector<detail::command_record> m_recorded_commands;
 };
 
+// Instruction recording
+
+struct buffer_allocation_record {
+	detail::buffer_id buffer_id;
+	std::string buffer_name;
+	detail::box<3> box;
+
+	friend bool operator==(const buffer_allocation_record& lhs, const buffer_allocation_record& rhs) {
+		return lhs.buffer_id == rhs.buffer_id && lhs.buffer_name == rhs.buffer_name && lhs.box == rhs.box;
+	}
+	friend bool operator!=(const buffer_allocation_record& lhs, const buffer_allocation_record& rhs) { return !(lhs == rhs); }
+};
+
+enum class instruction_dependency_origin {
+	allocation_lifetime,
+	write_to_allocation,
+	read_from_allocation,
+	side_effect,
+	collective_group_order,
+	last_epoch,
+	execution_front,
+	split_receive,
+};
+
+struct instruction_dependency_record {
+	instruction_id predecessor;
+	instruction_id successor;
+	instruction_dependency_origin origin;
+
+	instruction_dependency_record(const instruction_id predecessor, const instruction_id successor, const instruction_dependency_origin origin)
+	    : predecessor(predecessor), successor(successor), origin(origin) {}
+};
+
+/// IDAG base record type for `detail::instruction`.
+struct instruction_record
+    : matchbox::acceptor<struct clone_collective_group_instruction_record, struct alloc_instruction_record, struct free_instruction_record,
+          struct copy_instruction_record, struct device_kernel_instruction_record, struct host_task_instruction_record, struct send_instruction_record,
+          struct receive_instruction_record, struct split_receive_instruction_record, struct await_receive_instruction_record,
+          struct gather_receive_instruction_record, struct fill_identity_instruction_record, struct reduce_instruction_record, struct fence_instruction_record,
+          struct destroy_host_object_instruction_record, struct horizon_instruction_record,
+          struct epoch_instruction_record> //
+{
```

Comment:
IMO an approach with separate record types is superior because the consumer side knows which attributes are present from the type alone, and without having to inspect optionals. This is especially relevant for the IDAG, because the graph tests are formulated in terms of records as well.

I did consider using a `std::variant` instead of an inheritance hierarchy as well - would remove the heap allocation / `unique_ptr`s around records, but would mean that the common members like `id` would have to be replicated in every record type.